### PR TITLE
Add agent-based installer concept

### PR DESCRIPTION
**Pull request by zaneb**
The agent-based installer is an installation method that will use a bootable, ephemeral installer image running on the hosts that are to become part of the cluster. The user will generate the image using a command-line tool. The image itself will contain components (such as [assisted-service](https://github.com/openshift/assisted-service/#readme) and [assisted-agent](https://github.com/openshift/assisted-installer-agent#readme)) of the Assisted and Multi-Cluster Engine (MCE) installation methods. The installer will be usable in both a fully-automated workflow, where configuration is provided upfront, and an interactive workflow where the user can interact with a GUI similar to the hosted Assisted installation service.

### CODE REVIEW

File: enhancements/agent-installer/agent-based-installer.md
Comment (dhellmann) [inline enhancements/agent-installer/agent-based-installer.md:28]: It's covered well in the rest of the doc but it would be good to mention in the summary that the current focus is on-prem platforms.

File: enhancements/agent-installer/agent-based-installer.md
Comment (dhellmann) [inline enhancements/agent-installer/agent-based-installer.md:65]: You could elaborate on this by pointing out that the reason we want to be agnostic is that we have customers who want to bring their own provisioning tools.

File: enhancements/agent-installer/agent-based-installer.md
Comment (dhellmann) [inline enhancements/agent-installer/agent-based-installer.md:78]: We have some future goals that could be called out as non-goals for this enhancement that would be covered by other enhancements later

- Support multiple image formats
- Create minimal bootable images and separate root filesystems

File: enhancements/agent-installer/agent-based-installer.md
Comment (dhellmann) [inline enhancements/agent-installer/agent-based-installer.md:100]: I still have questions about whether what's described here is the best UX we can provide. Let's put the interactive flow on the non-goals list for this document and focus on the automated flow.

File: enhancements/agent-installer/agent-based-installer.md
Comment (dhellmann) [inline enhancements/agent-installer/agent-based-installer.md:228]: What is the third boot, after booting the live ISO with the agent and then rebooting to run as a node?

File: enhancements/agent-installer/agent-based-installer.md
Comment (dhellmann) [inline enhancements/agent-installer/agent-based-installer.md:128]: This section would be a good place to explain the reasoning behind each of the decisions described in the Proposal section.

"We must be able to run without a service on a host outside of the cluster, therefore we will build images that are self-orchestrating."

"We must support user-provided provisioning tools, therefore we will produce artifacts like ISOs that are ready to be consumed by those tools with minimal integration."

"We must support user-provided manifests like the openshift-installer does and will not have a REST API service, therefore the inputs to the image generator tool will be files."

"We want to minimize the storage requirements for provisioning images, therefore we will build one image that can be used on all hosts instead of creating host-specific images."

File: enhancements/agent-installer/agent-based-installer.md
Comment (dhellmann) [inline enhancements/agent-installer/agent-based-installer.md:91]: There's also an open question about this point below. Maybe here just say that the images will be pulled from the same local registry?

File: enhancements/agent-installer/agent-based-installer.md
Comment (dhellmann) [inline enhancements/agent-installer/agent-based-installer.md:190]: This section could talk about the goal of supporting versionable manifests as inputs so that the API of the image generator follows our existing support and deprecation guarantees for anyone integrating the tool with their workflow.

File: enhancements/agent-installer/agent-based-installer.md
Comment (zaneb) [reply enhancements/agent-installer/agent-based-installer.md:228]: 1. Boot with live ISO, install to disk
2. MCO updates ostree on disk to match machine OS content shipped in cluster
3. Boot as Node

**Comment by dhellmann** [comment]
> Thanks for creating this Enhancement PR and articulating the scope, approaches and decisions we are making. It may be useful to add a diagram of what the automated workflow looks like for the user.

+1, I have a slide I can give you as a basis for that, @zaneb 

> For goals, it may be worth adding it's targetting initial cluster (cluster 0) in fully-disconnected environments, so folks don't interpret this to mean any cluster (cluster 1, cluster 2, etc.)
> 
> It's unclear what the how or what input is provided in the statement "all configuration data for the cluster to be installed (including the expected number of hosts) will be provided at the time the ISO is generated."

If the team has done enough experimentation to feel comfortable making a decision about that, let's write it up as a separate enhancement.

> We should add some info about what the pre-reqs (mirrored registry), and also what happens after the cluster is stood up, specifically, what's the direction the user takes for day 2 activities even if it's to point to other tooling options to continue).

I think the mirror prereq is mentioned. Let's handle day 2 activities as another enhancement.

File: enhancements/agent-installer/agent-based-installer.md
Comment (romfreiman) [inline enhancements/agent-installer/agent-based-installer.md:51]: It does available for dual stack and will be available to ipv6 once quay will be. 
I would limit it to disconnected?

File: enhancements/agent-installer/agent-based-installer.md
Comment (romfreiman) [inline enhancements/agent-installer/agent-based-installer.md:74]: :(

File: enhancements/agent-installer/agent-based-installer.md
Comment (romfreiman) [inline enhancements/agent-installer/agent-based-installer.md:75]: Why do you mind?

File: enhancements/agent-installer/agent-based-installer.md
Comment (romfreiman) [reply enhancements/agent-installer/agent-based-installer.md:78]: Why not? Isn't it built in already? Ipxe and minimal iso is there. I would say that releasing without minmal iso is sbuoptimal

Comment (romfreiman) [reply enhancements/agent-installer/agent-based-installer.md:78]: It's a good point. Is it acceptable releasing without minimal iso support? 
@julienlim

File: enhancements/agent-installer/agent-based-installer.md
Comment (avishayt) [inline enhancements/agent-installer/agent-based-installer.md:256]: This is unfortunate.  Having this would mean that we could install using the same code/logic as expanding a cluster (via MAPI/CAPI).

File: enhancements/agent-installer/agent-based-installer.md
Comment (zaneb) [reply enhancements/agent-installer/agent-based-installer.md:74]: @romfreiman why would that be a goal?

File: enhancements/agent-installer/agent-based-installer.md
Comment (zaneb) [reply enhancements/agent-installer/agent-based-installer.md:78]: AIUI users can build a minimal ISO from the ISO using coreos-installer, and they can certainly build PXE artifacts. We might want to document this but I don't think we have to support it directly in the tool to have a viable installer.

Comment (dhellmann) [reply enhancements/agent-installer/agent-based-installer.md:78]: Right, remember this doc isn't the only place we'll talk about work on this project. Right now we're not focusing on multiple image formats. We know we'll need to, and we'll want a separate document when we do.

File: enhancements/agent-installer/agent-based-installer.md
Comment (dhellmann) [reply enhancements/agent-installer/agent-based-installer.md:74]: This document does not describe how we would run an API service that an orchestration tool would use to trigger provisioning because we have requirements that have led us to conclude that the tool can't work that way. Some of the details I've suggested adding in my comment below (https://github.com/openshift/enhancements/pull/1053/files#r821861322) will make that clearer.

It's possible this bit of information would fit better in the proposal or details section (maybe in the rough overview of the workflow that @julienlim asked about), since it's a conclusion rather than a starting condition for the design.

Comment (zaneb) [reply enhancements/agent-installer/agent-based-installer.md:74]: I included it in the non-goals because it isn't a goal: no user has requested it and nobody afaik has articulated a need for it. If it _were_ a goal then we would need to adjust the design and/or compromise on other goals. Nevertheless, it comes up a lot so it is worth being explicit about this as a starting condition.

This doesn't necessarily mean that it is an anti-goal, but the _conclusion_ is that due to the other constraints we have we're not going to have one.

Comment (romfreiman) [reply enhancements/agent-installer/agent-based-installer.md:74]: Even just because we're in 2022 and everything in the world is API driven

File: enhancements/agent-installer/agent-based-installer.md
Comment (romfreiman) [inline enhancements/agent-installer/agent-based-installer.md:57]: I assume progress monitoring is one of the goals as well, right?

File: enhancements/agent-installer/agent-based-installer.md
Comment (romfreiman) [inline enhancements/agent-installer/agent-based-installer.md:130]: I think installation progress reporting is a risk, no?

File: enhancements/agent-installer/agent-based-installer.md
Comment (romfreiman) [reply enhancements/agent-installer/agent-based-installer.md:228]: hmm are u sure about the 3rd one?

File: enhancements/agent-installer/agent-based-installer.md
Comment (romfreiman) [inline enhancements/agent-installer/agent-based-installer.md:263]: They still mirror acm, operators...

File: enhancements/agent-installer/agent-based-installer.md
Comment (romfreiman) [inline enhancements/agent-installer/agent-based-installer.md:266]: Theoretically, and practiallcaly, u can always use the same iso as a discovery entry point - anyhow the os-content is overwritten by MCO and there are no optimizations there. 
Few challenges are ignition version (that might be not backwards compatible) and drivers difference between the live iso and rhcos node os

File: enhancements/agent-installer/agent-based-installer.md
Comment (romfreiman) [inline enhancements/agent-installer/agent-based-installer.md:59]: Isnt multi version support is a requirement?

File: enhancements/agent-installer/agent-based-installer.md
Comment (zaneb) [reply enhancements/agent-installer/agent-based-installer.md:228]: @cgwalters can you confirm/deny?

File: enhancements/agent-installer/agent-based-installer.md
Comment (zaneb) [reply enhancements/agent-installer/agent-based-installer.md:59]: Nominally it isn't, but I'll get back to you in about 90 minutes :smile:

File: enhancements/agent-installer/agent-based-installer.md
Comment (zaneb) [reply enhancements/agent-installer/agent-based-installer.md:266]: I agree, but we have to get the ISO from somewhere. Currently it's in the release payload, so that's the obvious place because there are no extra steps for the user.

File: enhancements/agent-installer/agent-based-installer.md
Comment (cgwalters) [reply enhancements/agent-installer/agent-based-installer.md:228]: That's correct, 3 boots today.  If condensing boots 2 and 3 is important, it would be quite viable to do an "offline" update of the metal image before booting it.  There's some risks to this (e.g. suddenly we'd actually be using a potentially new Ignition version) but I think we can generally cover it in CI.

File: enhancements/agent-installer/agent-based-installer.md
Comment (dhellmann) [reply enhancements/agent-installer/agent-based-installer.md:68]: Probably not the initial version because our first user isn't on ARM. It's on the roadmap for a near-future version.

**Comment by zaneb** [comment]
I've updated with most of the suggestions from comments. The remaining ones are mostly requesting more detail about the automated flow - that's valid, but I think it will be better to put it all into a dedicated enhancement so that all the info about the automated flow is together. In this one I'll keep the automated and interactive flows at the same (low) level of detail. I think we are close to being able to merge this after another round of reviews.

File: enhancements/agent-installer/agent-based-installer.md
Comment (imiller0) [inline enhancements/agent-installer/agent-based-installer.md:106]: Will the input support the full range of what is possible with current Assisted Service CRs (ClusterDeployment, AgentClusterInstall, NMStateConfig, InfraEnv and extra installation manifests)?

File: enhancements/agent-installer/agent-based-installer.md
Comment (zaneb) [reply enhancements/agent-installer/agent-based-installer.md:55]: I agree with you on the importance of automation, but that's not really the point here - the point is which parts of the automation the _user_ has control over.
If you want OpenShift to provide complete automation then IPI is available.
If you want to automate booting the hosts yourself and have OpenShift automate everything else, then the agent-based installer is for you, and if you are in a disconnected environment it is the only game in town.

**Comment by dhellmann** [comment]
> I think we are close to being able to merge this after another round of reviews.

Yes, I agree. Let's give reviewers a last chance to comment and then I'll approve it Wednesday afternoon if there is no  feedback requiring significant changes.

/cc @avishayt @celebdor @hardys @lranjbar @mhrivnak @patrickdillon @pawanpinjarkar @rwsu @sdodson

File: enhancements/agent-installer/agent-based-installer.md
Comment (imiller0) [reply enhancements/agent-installer/agent-based-installer.md:106]: Great. Thanks!

File: enhancements/agent-installer/agent-based-installer.md
Comment (imiller0) [reply enhancements/agent-installer/agent-based-installer.md:55]: I think I'm well aligned with what both of you are saying. My intent was that there is significant value in being able to drive this installer within an automated framework where the user has a set of inputs (declarative CRs) which result in the ISO.

Comment (zaneb) [reply enhancements/agent-installer/agent-based-installer.md:55]: Added this as a goal as suggested below.

File: enhancements/agent-installer/agent-based-installer.md
Comment (mhrivnak) [inline enhancements/agent-installer/agent-based-installer.md:49]: typo here. `s/users to //`

**Comment by patrickdillon** [comment]
Relationship to `openshift-install` is stated as an open question, which makes sense as we continue to work through that design. So this LGTM as it stands.

File: enhancements/agent-installer/agent-based-installer.md
Comment (mhrivnak) [inline enhancements/agent-installer/agent-based-installer.md:92]: I think this paragraph would be easier to follow if it were more linear.
* An openshift release needs to be available in a local mirror
* The tool gets the base RHCOS image from that release
* The tool creates a new ISO
* The new ISO gets booted (fixes some current ambiguity on line 95 about which is "this" ISO that gets booted)
* Other required container images get pulled from the registry after boot

**Review by mhrivnak — COMMENTED** [review]
Generally this looks good to me. It would be good to see the comment threads, and questions asked within, resolved before merging.

File: enhancements/agent-installer/agent-based-installer.md
Comment (dhellmann) [reply enhancements/agent-installer/agent-based-installer.md:241]: Let's leave this for the detailed enhancement to be written next.

File: enhancements/agent-installer/agent-based-installer.md
Comment (dhellmann) [reply enhancements/agent-installer/agent-based-installer.md:307]: It is, but that's well outside of the scope of work for this team.

File: enhancements/agent-installer/agent-based-installer.md
Comment (dhellmann) [reply enhancements/agent-installer/agent-based-installer.md:57]: We'll get into details about that in the next enhancement.

File: enhancements/agent-installer/agent-based-installer.md
Comment (rwsu) [inline enhancements/agent-installer/agent-based-installer.md:117]: Is the API the assisted-service REST-API?

**Review by rwsu — CHANGES_REQUESTED** [review]
Looks good to me. Requested a minor clarification.

**Review by cgwalters — APPROVED** [review]

File: enhancements/agent-installer/agent-based-installer.md
Comment (dhellmann) [reply enhancements/agent-installer/agent-based-installer.md:117]: Yes, that's right.

**Comment by avishayt** [comment]
LGTM, thanks

**Comment by dhellmann** [comment]
I think we have achieved consensus on the general approach. We'll be creating some more low-level enhancements very soon, so there will be more opportunities to ask questions and dig into details there.

/approve
/lgtm

File: enhancements/agent-installer/agent-based-installer.md
Comment (zaneb) [reply enhancements/agent-installer/agent-based-installer.md:117]: Probably, but the vagueness was intentional (we may rule out e.g. a k8s API running on node0 and started prior to bootstrapping, but not in this enhancement).

**Pull request merged**
